src/Microsoft.IdentityModel.Tokens/MaxCapacityRefreshCache.cs

```diff
@@ -0,0 +1,137 @@
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MaxCapacityRefreshCache{TKey,TValue}" /> class.
+        /// </summary>
+        /// <param name="sizeLimit">The max number of items the cache is capable of holding.</param>
+        /// <param name="comparer">The string comparer.</param>
+        internal MaxCapacityRefreshCache(
+            long sizeLimit,
+            IEqualityComparer<TKey> comparer = null)
+        {
+            RecycleLock = new object();
+            SizeLimit = sizeLimit;
+            Cache = new ConcurrentDictionary<TKey, TValue>(comparer ?? EqualityComparer<TKey>.Default);
+        }
+
+        internal bool Contains(TKey key)
+        {
+            return Cache.ContainsKey(key);
+        }
+
+        /// <summary>
+        /// Gets the value corresponding to <paramref name="value"/> from the cache.
+        /// </summary>
+        /// <param name="cacheKey">The cache key.</param>
+        /// <param name="value">The cached value.</param>
+        /// <remarks>This method is non-blocking and thread-safe.</remarks>
+        internal bool TryGetValue(TKey cacheKey, out TValue value)
+        {
+            RecycleCacheIfNeeded();
+            return Cache.TryGetValue(cacheKey, out value);           
+        }
+
+        /// <summary>
+        /// Sets the cache value.
+        /// </summary>
+        /// <param name="cacheKey">The cache key.</param>
+        /// <param name="value">The value.</param>
+        /// <returns><paramref name="value"/> if it was successfully set, null otherwise.</returns>
+        /// <remarks>This method is non-blocking and thread-safe.</remarks>
+        /// <remarks>This value will not be set if the cache already contains a key with the same name as <paramref name="cacheKey"/>.</remarks>
+        internal TValue SetValue(TKey cacheKey, TValue value)
+        {
+            RecycleCacheIfNeeded();
+            if (Cache.TryAdd(cacheKey, value))
+                return value;
+            else
+                return null;
+        }
+
+        /// <summary>
+        /// Removes a particular key from the cache.
+        /// </summary>
+        /// <param name="cacheKey">The cache key.</param>
+        internal bool TryRemove(TKey cacheKey)
+        {
+            RecycleCacheIfNeeded();
+            return Cache.TryRemove(cacheKey, out _);
+        }
+
+        /// <summary>
+        /// Removes the value corresponding to <paramref name="cacheKey"/>.
+        /// </summary>
+        /// <param name="cacheKey">The cache key.</param>
+        /// <param name="removedValue">The removed value.</param>
+        internal bool TryRemove(TKey cacheKey, out TValue removedValue)
+        {
+            RecycleCacheIfNeeded();
+            return Cache.TryRemove(cacheKey, out removedValue);
+        }
+
+        /// <summary>
+        /// Recycles the cache if needed.
+        /// </summary>
+        private void RecycleCacheIfNeeded()
```

Comment:
Do we need any read after write guarantees on this cache? I think this behavior would be surprising to most users. I think a simple LRU implementation would be better in most cases. Thoughts?

Reply:
+1, if we want to keep the Cache behavior consistent across different targets.
That said, for my knowledge I'd like to know the reason why a more intelligent solution is not worth it for net45.

Reply:
That's a valid point. I think my current approach only works if we operate under the assumption that the net45 target is going to be rarely used.

It's also worth taking into account that no one has actually run into any kind of overflow error with the InMemoryCryptoProviderCache, so the work that's currently being done here is more of a preventative measure. Even this simple approach is better than what we're currently doing.

A simple LRU implementation shouldn't be too difficult to implement, so I can modify the cache to do that instead if we believe the extra complexity is necessary.